### DISCUSSION TITLE
Retrospective - using PocketBase for a realtime open world multiplayer web game

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#540] [Q&A]
I am working on a web game with providers implemented in both Supabase and Firebase. I decided to try adding a Pocketbase provider, and here's what came up. If anyone has solutions or ideas, I'm listening :)

## Anonymous users. 

My game allows users to play anonymously and then upgrade to a full user account as needed. 

In Firebase, anonymous logins get real user IDs. This seems ideal. 

In Supabase, I have to create a fake user and suppress the confirmation email. The user record is then created, and if they ever want to upgrade to a real email address I have to figure out how to change it. Less than ideal. 

In PocketBase, I'm not sure what to do. Currently, I'm creating a fake user account via `authViaEmail` and keeping it in `localStorage` so I can have a somewhat persistent session that outlives the auth token. I think the ideal solution would be to have a `client.authAnonymously` that would create a real user record that could be persistent and eventually upgraded with an email/pass if desired. Perhaps the anonymous user accounts could be garbage-collected periodically if the dev wanted to.

## Realtime presence and disconnect

Covered in #467 and #468. Firebase has a great feature called `onDisconnect` where you can send a mutation ahead of time that will be run when the socket disconnects for any reason. If PocketBase had a `OnRealtimeDisconnectHook` event, it could be customize to do that work. But what would be even better is a `client.realtime.onDisconnect` that works like this:

```ts
const cancelFn = await client.realtime.onDisconnect({
  'myCollection/myId': { foo: 42 },
  'myOtherCollection/someOtherId': { bar: 'baz' }
}).catch(console.error)

// If I want to cancel the queued action
cancelFn().catch(console.error)
```

`onDisconnect` would send that payload immediately and PocketBase would try it in a transaction to make sure it works (and then cancel the transaction). If the test transaction fails, the client throws an error. Otherwise, it returns a function to cancel the queued action. 

If it could accept a hash of basic updates to perform, PocketBase could queue that in memory and fire it when the user's socket connection drops.

If PocketBase had a `OnRealtimeDisconnectHook` event, this could probably be implemented using that hook plus an `realtime.onDisconnect` API call. It would obviously be nice to have first-class support in the client libs.

## Realtime Collection filtering

My game is an unbounded grid of tiles. Depending on the player's position in the world, it would be great to have realtime watches on roughly a 10-cell radius from the player's position. 

Firebase's realtime database can handle up to 100 watches per client. By grouping the cells into 10x10 chunks, I can further reduce the number of watchers needed (at the expense of getting unwanted cell updates). This solution seems close to ideal, and it works.

Supabase's realtime features are too new to tell. I can watch specific rows by field value using a filter syntax (`id=eq.23`). As long as I can do that on an arbitrary field and not just the primary key field (I haven't tested), I should be able to create watches on the cell blocks just like I can in Firebase. If that isn't possible, I have to resort to polling. 

PocketBase seems the most limited. I could just poll, and maybe I will. But if I did want to use the realtime capability of PocketBase, the best I can think of is this:

1. Create a `cells` collection with `cellBlockId` and `updateUuid (unique)`
2. Create a `realtimeCells` collection with `cellBlockId (unique)`, `cellId`, `updateUuid (unique)`.
3. Whenever I update `cells`, I tag the record with an `updateUuid` and update the corresponding `realtimeCells.cellBlockId` with `cellId = cells.id` and `updateUuid = cells.updateUuid`
4. When I get a realtime update, compare the `updateUuid` field to the current cell in memory. If they don't match, I re-query that cell block

All of that seems like a lot of work vs just polling or realtime filtering. How feasible is it to implement realtime filters?

## Hosting (PocketHost.io)

Firebase and Supabase both have BaaS managed services. It's super nice for hobby projects to just sign up and get a project URL.

I realize PocketBase really has self-hosting in mind, but that wasn't working well for me. @ganigeorgiev hadn't written #537 yet (thank you!!!) and anyway I was having a heck of a time getting a Dockerized PocketBase deployed. I was stuck on localhost. And even if I did get it deployed, I would still have email sending to worry about if I start allowing registered users, plus the ongoing task of maintaining the Docker image, updates, etc. I really didn't want to manage a backend. Supabase and Firebase have spoiled me.

So, I took the weekend to create [PocketHost.io](https://pockethost.io) (free) where anyone can spin up a PocketBase instance with one button push, zero config, and zero ongoing management. Writing the hosting service familiarized me a lot with PocketBase, so overall it was time well spent.

**Comment by ganigeorgiev** [comment]
1. > Anonymous users.

    PocketBase doesn't really need "anonymous users" because both the data operations and the realtime events are based on the collection API rules.

    With the refactoring, we will have support for login via username (which by default will be auto generated), so if you want to associate some data with your "anonymous users", you can create an account with username (aka. similar to your approach, but without the fake email workaround).

2. > Realtime presence and disconnect

    This is a little more complex, because SSE has auto reconnect by design and allowing sending mutation on every disconnect attempt may not be expected by everyone. I'll consider adding in the future a "disconnect" client-side callback, which will allow users to specify whether they want the reconnect functionality or not.


3. > Realtime Collection filtering

    I'm not sure that I fully understand your use case and why you need 2 collections (`cells` and `realtimeCells`). You can subscribe to an entire collection, eg. cells and you can further filter the event record with some local state check, eg:
    ```js
    client.realtime.subscribe('cells', function (e) {
        if (e.record.cellBlockId === "test") {
            // do something with the record...
            console.log(e.record);
        }
    });
    ```
    When you subscribe to an entire collections, the "list" API rule will be used to determine whether the connected client is allowed to receive events happening in that collection.
    If you subscribe to a single record (eg. "cells/record_id") - the "view" API rule will be used.

    For subscriptions filtering and queries you can watch this issue - https://github.com/pocketbase/pocketbase/issues/527#issuecomment-1250155083.

**Reply by benallfree** [reply]
> why you need 2 collections

Sitting on the full `cells` realtime feed could get very busy. But yes, client-side filtering is always an option :)

**Reply by pchasco** [reply]
> 1. > Anonymous users.
>    
>    
>    PocketBase doesn't really need "anonymous users" because both the data operations and the realtime events are based on the collection API rules.
>    With the refactoring, we will have support for login via username (which by default will be auto generated), so if you want to associate some data with your "anonymous users", you can create an account with username (aka. similar to your approach, but without the fake email workaround).

In this scenario, how would one link this anonymous user to an OAuth user, such as they do in Firebase? This is very important functionality for apps that do not want to immediately ask users to sign up, but do need to store data in the server for that user. The example use case cited by Firebase is a shopping cart where users have already added items to a cart prior to sign-in.

**Reply by ganigeorgiev** [reply]
> how would one link this anonymous user to an OAuth user

You can link the OAuth2 account with the anonymous user as long as the PocketBase user is already authenticated and the OAuth2 account is not linked to another PocketBase user. In other words:

```js
// authenticate the created anonymous user in the background
await pb.collection("users").authWithPassword("somerandomusername", "password")

// authenticate with OAuth2
// (will send the previously authenticated user token with the OAuth2 auth request,
// which should link the OAuth2 acount, if not already linked to another user)
await pb.collection("users").authWithOAuth2({ provider: "google" });
```

**Reply by mukhtharcm** [reply]
What if I want to link to a real email instead?

**Reply by ganigeorgiev** [reply]
@mukhtharcm I'm not sure that I understand what do you mean by "_link to a real email instead_".
We can't allow public OAuth2 linking by any random email address because it will be a security issue, if that is what you want.
You can link the OAuth2 user to any auth record manually if you use for example the [`OnRecordBeforeAuthWithOAuth2Request` hook](https://pocketbase.io/docs/go-event-hooks/#onrecordbeforeauthwithoauth2request) by assigning `e.Record` to your desired auth record model.
If the above doesn't help, please open new discussion with more details about your use case.

**Reply by mukhtharcm** [reply]
Thank you for this awesome library man!
After reading and experimenting, I feel like that question was useless.

my current workflow is,
on app startup, if there's no account logged in, create an account with random username and a random password. and store them in localstorage

if the user choose to create an account, get email and password from them, then update the user record with our previousely stored password.

Again thanks :)
